examples/Cassie/cassie_rbt_state_estimator.cc

```diff
@@ -1,140 +1,981 @@
+
+  // Initialize body indices
+  left_thigh_ind_ = GetBodyIndexFromName(tree, "thigh_left");
+  right_thigh_ind_ = GetBodyIndexFromName(tree, "thigh_right");
+  left_heel_spring_ind_ = GetBodyIndexFromName(tree, "heel_spring_left");
+  right_heel_spring_ind_ = GetBodyIndexFromName(tree, "heel_spring_right");
+  if (left_thigh_ind_ == -1 || right_thigh_ind_ == -1 ||
+      left_heel_spring_ind_ == -1 || right_heel_spring_ind_ == -1 )
+    std::cout << "In cassie_rbt_state_estimator.cc,"
+              " body indices were not set correctly.\n";
+
+  if (is_floating_base) {
+    // Declare input port receiving robot's state (simulation ground truth state)
+    // TODO(yminchen): delete this input port after finishing testing
+    state_input_port_ = this->DeclareVectorInputPort(
+                          OutputVector<double>(tree_.get_num_positions(),
+                              tree_.get_num_velocities(),
+                              tree_.get_num_actuators())).get_index();
+
+    // Declare update event for EKF
+    DeclarePerStepDiscreteUpdateEvent(&CassieRbtStateEstimator::Update);
+
+    // a state which stores previous timestamp
+    time_idx_ = DeclareDiscreteState(VectorXd::Zero(1));
+
+    // states related to EKF
+    VectorXd init_floating_base_state = VectorXd::Zero(7 + 6);
+    init_floating_base_state(3) = 1;
+    state_idx_ = DeclareDiscreteState(
+        init_floating_base_state);  // estimated floating base
+    ekf_X_idx_ = DeclareDiscreteState(27);  // estimated EKF state
+
+    // states related to contact estimation
+    previous_velocity_idx_ = DeclareDiscreteState(
+        VectorXd::Zero(tree_.get_num_velocities(), 1));
+
+    filtered_residue_double_idx_ = DeclareDiscreteState(
+        VectorXd::Zero(tree_.get_num_velocities(), 1));
+    filtered_residue_left_idx_ = DeclareDiscreteState(
+        VectorXd::Zero(tree_.get_num_velocities(), 1));
+    filtered_residue_right_idx_ = DeclareDiscreteState(
+        VectorXd::Zero(tree_.get_num_velocities(), 1));
+
+    ddq_double_init_idx_ = DeclareDiscreteState(
+        VectorXd::Zero(tree_.get_num_velocities(), 1));
+    ddq_left_init_idx_ = DeclareDiscreteState(
+        VectorXd::Zero(tree_.get_num_velocities(), 1));
+    ddq_right_init_idx_ = DeclareDiscreteState(
+        VectorXd::Zero(tree_.get_num_velocities(), 1));
+    lambda_b_double_init_idx_ = DeclareDiscreteState(VectorXd::Zero(2, 1));
+    lambda_b_left_init_idx_ = DeclareDiscreteState(VectorXd::Zero(2, 1));
+    lambda_b_right_init_idx_ = DeclareDiscreteState(VectorXd::Zero(2, 1));
+    lambda_cl_double_init_idx_ = DeclareDiscreteState(VectorXd::Zero(6, 1));
+    lambda_cl_left_init_idx_ = DeclareDiscreteState(VectorXd::Zero(6, 1));
+    lambda_cr_double_init_idx_ = DeclareDiscreteState(VectorXd::Zero(6, 1));
+    lambda_cr_right_init_idx_ = DeclareDiscreteState(VectorXd::Zero(6, 1));
+  }
+}
+
+void CassieRbtStateEstimator::solveFourbarLinkage(
```

Comment:
This function is going to be hard for anyone to understand/debug. It could use a *lengthy* block of documentation describing how it works.

Reply:
Done!